Android/app/src/main/java/com/peer1/internetmap/utils/TracerouteUtil.java

```diff
@@ -0,0 +1,198 @@
+package com.peer1.internetmap.utils;
+
+import android.os.AsyncTask;
+import android.util.Log;
+
+import com.peer1.internetmap.MapControllerWrapper;
+import com.peer1.internetmap.ProbeWrapper;
+
+import java.util.HashSet;
+
+/**
+ * Created by shayla on 2017-09-20.
+ *
+ * Simulates a traceroute by requesting probes be sent to a given destination IP with incrementing
+ * time-to-live values. Reports back via the Listener interface.
+ */
+
+public class TracerouteUtil {
+
+    //=======================================================================
+    // Callback interface; must be set by caller to receive traceroute updates
+    //=======================================================================
+    public interface Listener {
+        /**
+         * Successful probe of a given time-to-live hop
+         */
+        void onHopFound(int ttl, ProbeWrapper hop);
+
+        /**
+         * Unsuccessful probe of a given time-to-live hop
+         */
+        void onHopTimeout(int ttl);
+
+        /**
+         * Traceroute completed successfully
+         */
+        void onComplete();
+
+        /**
+         * Traceroute exited early due to a loop being found in the trace (ie. duplicate IPs)
+         */
+        void onLoopDiscovered();
+
+        /**
+         * Traceroute timed out (usually due to consecutive hop timeouts)
+         */
+        void onTraceTimeout();
+    }
+
+    //=======================================================================
+    // Privates
+    //=======================================================================
+    private TracerouteTask tracerouteTask;
+
+    private Listener emptyListener = new Listener() {
+        @Override
+        public void onHopFound(int ttl, ProbeWrapper ip) { }
+
+        @Override
+        public void onHopTimeout(int ttl) { }
+
+        @Override
+        public void onComplete() { }
+
+        @Override
+        public void onTraceTimeout() { }
+
+        @Override
+        public void onLoopDiscovered() { }
+    };
+
+    //=======================================================================
+    // Constructors
+    //=======================================================================
+    public TracerouteUtil(MapControllerWrapper mapControllerWrapper) {
+        this.tracerouteTask = new TracerouteTask();
```

Comment:
The allocation of an independent task for each traceroute does worry me a little, unlike the rest of the MapControllerWrapper interface, I'm pretty sure the traceroute related stuff is not thread safe, so if there's ever a situation (probably a bug) where two traceroutes got run at the same time, bad things could happen. 

I think ideally this should be using some sort of queuing model, where it tries to ensure the there is only ever one traceroute task running at the same time, like a serial queue in GCD on iOS. Not en expert on Android threading, but I THINK the Android equivalent would be a ThreadPoolExecutor with the max thread count set to 1.

Reply:
Sorry, I don't fully understand how running 2 traceroutes at the same time would be an issue (I mean, from our current UI perspective yes, but from a threading standpoint).  Before I attempt a refactor, could you explain a little more how this theading issue could occur?

Reply:
Deleted last comment

Reply:
Given that I am attempting to reduce memory allocations by only using a single probeWrapper instance, I can see how this could lead to potential threading issues if multiple traceroutes are running at the same time.

Reply:
Since I wanted to be able to give feedback if a traceroute was still running I have updated TracerouteUtil to a singleton and exposed the running state of the traceroute. Only one traceroute should be able to run at a given time now.